Add drafts for RFC 6 & 7 (Ethereum defs and Ether HTLC) (#38)
comit-network/RFCs: RFC-007-SWAP-Ethereum-Basic.md

## Hash Time Lock Contract

### Hash Functions

This RFC specifies SHA-256 is the only value the `hash_function` header may take if Ethereum is used as a ledger.
This may be expanded in subsequent RFCs.

### Parameters

The parameters for the Ether HTLC follow [RFC003](./RFC-003-SWAP-basic.md) and are described concretely in the following table:

| Variable        | Description                                                                 |
|:----------------|:----------------------------------------------------------------------------|
| asset           | The quantity of wei                                                         |
| secret_hash     | The hash of the `secret` (32 bytes for SHA-256)                             |
| redeem_identity | The `address` of the redeeming party                                        |
| refund_identity | The `address` of the refunding party                                        |
| expiry          | The absolute UNIX timestamp in seconds after which the HTLC can be refunded |

### Contract

This RFC defines a single disposable contract as the HTLC.
It uses `selfdestruct` to release the funds to the intended party.

> This approach was chosen over a stateful contract written in solidity so the contracts can be precisely defined, verified and updated.

Reviewer: ```suggestion
This approach was chosen over a stateful contract written in Solidity to avoid having to deal with topics like upgradability or run into the risk of endangering the funds of multiple users due to a bug in the contract.
```

Author: How does not using a stateful contract not endanger the funds of multiple users due to a bug in the contract though?

The contract itself works by hard coding the `redeem_identity` and `refund_identity` into the contract.

> When it is called with the correct `secret` as the calldata it will transfer all the Ether to the `redeem_identity` using a `selfdestruct`.

Reviewer: ```suggestion
When it is called with the correct `secret` as the calldata it will transfer all the Ether to the `redeem_identity` using `selfdestruct`.
```

> When it is called without any calldata after the `expiry` it will transfer all the Ether to the `refund_identity` using a `selfdestruct`.

Reviewer: ```suggestion
When it is called without any calldata after the `expiry` it will transfer all the Ether to the `refund_identity` using `selfdestruct`.
```

> The contract never checks who is calling the contract so it can be called successfully by anyone.

Reviewer: ```suggestion
The contract itself actually never checks for the caller so it can be successfully called by practically anyone.
```

Author: I simplified this sentence to:
>The contract never checks the caller so the redeem and refund transactions can be sent by anyone.

    exit
    jumpi

    // Load secret into memory
    calldatacopy(0, 0, 32)

    // Hash secret with SHA-256 (pre-compiled contract 0x02)
    call(72, 0x02, 0, 0, 32, 33, 32)

    // Placeholder for correct secret hash
    <secret_hash>

    // Load hashed secret from memory
    mload(33)

    // Compare hashed secret with existing one
    eq

    // Combine `eq` result with `call` result
    and

    // Jump to redeem if hashes match
    redeem
    jumpi

    // Exit if hashes don't match
    return(0, 0)

check_expiry:
    // Timestamp of the current block in seconds since the epoch
    timestamp

    // Placeholder for refund timestamp
    <expiry>

    // Compare refund timestamp with current timestamp
    lt

    // Jump to refund if time is expired
    refund
    jumpi

exit:
    // Exit
    return(0, 0)

redeem:
    log1(0, 32, 0xB8CAC300E37F03AD332E581DEA21B2F0B84EAAADC184A295FEF71E81F44A7413) // log keccak256("Redeemed()")
    selfdestruct(<redeem_identity>)

refund:
    log1(0, 0, 0x5D26862916391BF49478B2F5103B0720A842B45EF145A268F2CD1FB2AED55178) // log keccak256("Refunded()")
    selfdestruct(<refund_identity>)
}

```

> The following is the contract deployment code encoded template as hex which will deploy the above contract when the placeholder values (set to `00`s) are replaced with the parameters.

Reviewer: Maybe we should leave the `<placeholder>` in there like that since it is totally valid for `00` to appear in the contract?

Author: I like the idea. The placeholders have to be the right length obviously.
The thing is I wouldn't expect them to store the placeholders as hex but as bytes and overwrite the bytes (and there's no such thing as an invalid byte sequence!)

| `Refunded()` | 0x5d26862916391bf49478b2f5103b0720a842b45ef145a268f2cd1fb2aed55178 | -        | The contract is refunded |

Implementations SHOULD use the following gas limits on the transactions related to the contract:

| Transaction | recommended gas limit |
|:------------|-----------------------|
| Deployment  | 121,000               |
| Redeem      | 100,000               |
| Refund      | 100,000               |

## Execution Phase

The following section describes how both parties should interact with the Ethereum blockchain during the [RFC003 execution phase](./RFC-003-SWAP-basic.md#execution-phase).

### Deployment

At the start of the deployment stage, both parties compile the contract code as described in the previous section.
We will call this value `contract_code`.

To deploy the Ether HTLC, the *funder* must deploy the `contract_code`.
They SHOULD do this by sending a contract deployment transaction to the relevant Ethereum blockchain with:

- `contract_code` as the `data` of the transaction
- The quantity of wei specified in the Ether Asset header as the `value` of the transaction

> The funder SHOULD NOT do this executing the `CREATE` opcode from another contract.

Reviewer: Can you clarify what you mean with the sentence?

Author: Note @bonomat made a suggestion clarifying this which I took.

To be notified of the deployment event, both parties MAY watch the blockchain for a transaction with the `contract_code` as the data.
Upon observing the deployment transaction, both parties SHOULD record the address the contract was deployed to (referred to as `contract_address` from now on).

### Redeem

Before redeeming, the redeemer SHOULD wait until the deployment transaction has enough confirmation such that they consider it permanent.

> To redeem the HTLC, the redeemer SHOULD send a transaction to the `contract_address` with the `data` of the transaction set to the `secret`.

Reviewer: ```suggestion
To redeem the HTLC, the redeemer sends a transaction to the `contract_address` with the `data` of the transaction set to the `secret`.
```

Author: Why don't you want the SHOULD here? We're telling them how they SHOULD redeem it. They may redeem it by calling the contract from another contract.

To be notified of the redeem event, both parties SHOULD watch the blockchain for `contract_address` emitting the `Redeemed()` log.
If Ethereum is the `beta_ledger`, then the funder (Bob) MUST watch for such a log, extract the `secret` from the transaction receipt and continue the protocol.
In this case, Bob MUST NOT watch for a transaction sent to `contract_address` with the `secret` as `data`.

> If he does this, he may miss learning the `secret` by having the contract be redeemed by another contract.

Reviewer: ```suggestion
If he does this, he may miss learning the `secret` by having the contract be redeemed through another contract.
```
